interpreter/interpreter.go

```diff
@@ -151,6 +156,185 @@ func CompileRegexConstants(regexOptimizations ...*RegexOptimization) Interpretab
 	return decRegexOptimizer(regexOptimizations...)
 }
 
+type typeVerifier func(int64, ...*types.TypeValue) (bool, error)
+
+// CheckStringFormat checks the syntax and cardinality of any string.format calls present in the expression and reports
+// any errors at compile time.
+func CheckStringFormat(enableCheckArgTypes bool, verifier typeVerifier) InterpretableDecorator {
+	return func(inter Interpretable) (Interpretable, error) {
+		call, ok := inter.(InterpretableCall)
+		if !ok {
+			return inter, nil
+		}
+		if call.OverloadID() != "string_format" {
+			return inter, nil
+		}
+		args := call.Args()
+		if len(args) != 2 {
+			return nil, fmt.Errorf("wrong number of arguments to string.format (expected 2, got %d)", len(args))
+		}
+		fmtStrInter, ok := args[0].(InterpretableConst)
+		if !ok {
+			return inter, nil
+		}
+		fmtArgsInter, ok := args[1].(InterpretableConstructor)
+		if !ok {
+			return inter, nil
+		}
+		if fmtArgsInter.Type() != types.ListType {
+			return nil, fmt.Errorf("expected list as second argument to string.format (got %s", fmtArgsInter.Type().TypeName())
+		}
+		formatStr := fmtStrInter.Value().Value().(string)
+		formatArgs := fmtArgsInter.InitVals()
+		curArgIndex := 0
+		i := 0
+		for i < len(formatStr) {
```

Comment:
It looks like this is a duplication implementation of the lexer? Can a single lexer implementation be shared for both runtime evaluation and compile time checking?  I'm imagining a function that invokes a callback for each directive it encounters that could then be used in both places.

Reply:
Fixed, I've refactored so there's only one lexer now.